src/upchunk.ts

```diff
@@ -1,9 +1,169 @@
 import { EventTarget, Event } from 'event-target-shim';
-import xhr, { XhrUrlConfig, XhrHeaders, XhrResponse } from 'xhr';
+import xhr, { type XhrUrlConfig, type XhrHeaders, type XhrResponse } from 'xhr';
+
+const DEFAULT_CHUNK_SIZE = 30720;
+const DEFAULT_MAX_CHUNK_SIZE = 512000; // in kB
+const DEFAULT_MIN_CHUNK_SIZE = 256; // in kB
+
+// Predicate function that returns true if a given `chunkSize` is valid, otherwise false.
+export const isValidChunkSize = (
+  chunkSize: any,
+  {
+    minChunkSize = DEFAULT_MIN_CHUNK_SIZE,
+    maxChunkSize = DEFAULT_MAX_CHUNK_SIZE,
+  } = {}
+): chunkSize is number | null | undefined => {
+  return (
+    chunkSize == null ||
+    (typeof chunkSize === 'number' &&
+      chunkSize >= 256 &&
+      chunkSize % 256 === 0 &&
+      chunkSize >= minChunkSize &&
+      chunkSize <= maxChunkSize)
+  );
+};
+
+// Projection function that returns an error associated with invalid `chunkSize` values.
+export const getChunkSizeError = (
+  chunkSize: any,
+  {
+    minChunkSize = DEFAULT_MIN_CHUNK_SIZE,
+    maxChunkSize = DEFAULT_MAX_CHUNK_SIZE,
+  } = {}
+) => {
+  return new TypeError(
+    `chunkSize ${chunkSize} must be a positive number in multiples of 256, between ${minChunkSize} and ${maxChunkSize}`
+  );
+};
+
+export type ChunkedStreamIterableOptions = {
+  defaultChunkSize?: number;
+  minChunkSize?: number;
+  maxChunkSize?: number;
+};
+
+// An Iterable that accepts a readableStream of binary data (Blob | Uint8Array) and provides
+// an asyncIterator which yields Blob values of the current chunkSize until done. Note that
+// chunkSize may change between iterations.
+export class ChunkedStreamIterable implements AsyncIterable<Blob> {
+  protected defaultChunkSize: number;
+  public readonly minChunkSize: number;
+  public readonly maxChunkSize: number;
+
+  constructor(
+    protected readableStream: ReadableStream<Uint8Array | Blob>,
+    options: ChunkedStreamIterableOptions = {}
+  ) {
+    if (!isValidChunkSize(options.defaultChunkSize, options)) {
+      throw getChunkSizeError(options.defaultChunkSize, options);
+    }
+    this.defaultChunkSize = options.defaultChunkSize ?? DEFAULT_CHUNK_SIZE;
+    this.minChunkSize = options.minChunkSize ?? DEFAULT_MIN_CHUNK_SIZE;
+    this.maxChunkSize = options.maxChunkSize ?? DEFAULT_MAX_CHUNK_SIZE;
+  }
+
+  protected _chunkSize: number | undefined;
+
+  get chunkSize() {
+    return this._chunkSize ?? this.defaultChunkSize;
+  }
+
+  set chunkSize(value) {
+    if (!isValidChunkSize(value, this)) {
+      throw getChunkSizeError(value, this);
+    }
+    this._chunkSize = value;
+  }
+
+  get chunkByteSize() {
+    return this.chunkSize * 1024;
+  }
+
+  async *[Symbol.asyncIterator](): AsyncIterator<Blob> {
```

Comment:
what is upchunk's expected browser support matrix? Does our usage of Symbols, generators, etc. inadvertently change it? Or does it compile away with regenerator runtime?

Reply:
It's always been a bit squishy, but, given the hard dependency on `ReadableStream` (and the corresponding update to the File API) for this effort, we should be good. Good callout though.

Reply:
cool, and since you're targeting making this a major, it should be fine. (I also wasn't sure what the overlap between generators/ReadableStream/Symbols was, but probably large enough not to worry about except IE).
Worth solidifying the support matrix somewhere, though.